jgrapht-core/src/main/java/org/jgrapht/alg/treedecomposition/NiceDecompositionBuilder.java

```diff
@@ -0,0 +1,196 @@
+package org.jgrapht.alg.treedecomposition;
+
+import java.util.*;
+
+import org.jgrapht.*;
+import org.jgrapht.alg.util.*;
+import org.jgrapht.graph.*;
+
+/**
+ * A builder for nice tree decompositions. A tree decomposition of a graph G is a tree T and a map
+ * b: V(T) to Set of V(G), which satisfies the properties: - for every edge e in E(G), there is a
+ * node t in V(T) with e is a subset of b(v) - for all vertices v in V(G) the set {t in V(T) | v in
+ * b(t)} is non-empty and connected in T
+ * 
+ * A nice tree decomposition is a special tree decomposition, which satisfies the properties: - for
+ * root r in V(T) and leaf l in V(T): b(r)=b(t)=empty set - every non-leaf node t in V(T) is of one
+ * of the following three types: - introduce node: t has exactly one child d and b(t) = b(d) union w
+ * for some w in V(G) - forget node: t has exactly one child d and b(t) union w = b(d) for some w in
+ * V(G)\b(t) - join node: t has exactly two child d_1, d_2 and b(t)=b(d_1)=b(d_2)
+ * 
+ * @author Ira Justus Fesefeldt (PhoenixIra)
+ *
+ * @param <V> the vertices of G
+ */
+public class NiceDecompositionBuilder<V>
+{
+
+    // resulting forest of the decomposition
+    private Graph<Integer, DefaultEdge> decomposition;
+
+    // map from decomposition nodes to the interval sets
+    private Map<Integer, Set<V>> decompositionMap;
+
+    // the root of the tree
+    private Integer root;
+
+    // next integer for vertex generation
+    private Integer nextInteger;
+
+    /**
+     * constructor for all methods used in the abstract method
+     */
+    public NiceDecompositionBuilder()
+    {
+        // creating objects
+        decomposition = new DefaultDirectedGraph<Integer, DefaultEdge>(DefaultEdge.class);
+        decompositionMap = new HashMap<Integer, Set<V>>();
+
+        // create root
+        root = 0;
+        nextInteger = 1;
+        decompositionMap.put(root, new HashSet<V>());
+        decomposition.addVertex(root);
+    }
+
+    /**
+     * getter for the next free Integer
+     * 
+     * @return unused integer
+     */
+    private Integer getNextInteger()
```

Comment:
it is in essence a counter.

Reply:
It is implemented as a counter, but not used as a counter. It is used as a supplier. But using a Suppier Interface here seeme like a bit overkill to me...?